utils/driverless.c

```diff
@@ -673,19 +679,221 @@ list_printers (int mode ,int reg_type_no)
+    /* Mark all the fields of the output of ippfind */
+    ptr = buffer;
+    /* First, build the DNS-SD-service-name-based URI ... */
+      while (ptr && !isalnum(*ptr & 255)) ptr ++;
+      
+    service_hostname = ptr; 
+        
+    ptr = memchr(ptr, '\t', sizeof(buffer) - (ptr - buffer));
+    if (!ptr) goto read_error;
+    *ptr = '\0';
+    ptr ++;
+    ptr_to_port = ptr;
+    ptr = memchr(ptr, '\t', sizeof(buffer) - (ptr - buffer));
+    if (!ptr) goto read_error;
+    *ptr = '\0';
+    ptr ++;
+    port = convert_to_port(ptr_to_port);
+    httpAssembleURIf(HTTP_URI_CODING_ALL, service_uri,
+      2047,scheme, NULL,service_hostname, port, (isFax?"/ipp/faxout":"/ipp/print"));
+   
+  read_error:
+      continue;
+  }
+
+/*
+  * Wait for the child processes to exit...
+  */
+
+  wait_children = 1;
+
+  while (wait_children > 0){
+   /*
+    * Wait until we get a valid process ID or the job is canceled...
+    */
+
+    while ((wait_pid = wait(&wait_status)) < 0 && errno == EINTR){
+      if (job_canceled){
+      	kill(ippfind_pid, SIGTERM);
+	      job_canceled = 0;
+      }
+    }
+
+    if (wait_pid < 0)
+      break;
+
+    wait_children --;
+   /*
+    * Report child status...
+    */
+
+    if (wait_status){
+      if (WIFEXITED(wait_status)){
+	      exit_status = WEXITSTATUS(wait_status);
+
+        if (debug)
+          fprintf(stderr, "DEBUG: PID %d (%s) stopped with status %d!\n",
+          wait_pid,(wait_pid == ippfind_pid ? "ippfind" :"Unknown process"),exit_status);
+       
+        if (wait_pid == ippfind_pid && exit_status <= 2)
+          exit_status = 0;	  
+      }
+      else if (WTERMSIG(wait_status) == SIGTERM){
+        if(debug)
+          fprintf(stderr,"DEBUG: PID %d (%s) was terminated normally with signal %d!\n",
+            wait_pid,(wait_pid == ippfind_pid ? "ippfind" :"Unknown process"),exit_status);
+      }
+      else{
+	      exit_status = WTERMSIG(wait_status);
+        if (debug)
+          fprintf(stderr, "DEBUG: PID %d (%s) crashed on signal %d!\n",wait_pid,
+            (wait_pid == ippfind_pid ? "ippfind":"Unknown process"),exit_status);
+      }
+    }
+    else{
+      if (debug)
+	      fprintf(stderr, "DEBUG: PID %d (%s) exited with no errors.\n",wait_pid,
+		    (wait_pid == ippfind_pid ? "ippfind" :"Unknown process"));
+    }
+  }
+
+ /*
+  * Exit...
+  */
+
+  error:
+    return (exit_status);
+  }
 
 int
-generate_ppd (const char *uri)
-{
+generate_ppd ( char *uri ,int isFax){
+
   ipp_t *response = NULL;
   char buffer[65536], ppdname[1024];
-  int fd, bytes;
-  char *ptr1, *ptr2;
-
-  /* Request printer properties via IPP to generate a PPD file for the
+  int 
+    fd, 
+    bytes,
+    exit_status_uri_converter = 0;
+  char *ptr1, 
+       *ptr2,
+       *std_uri = NULL;
+   /* Request printer properties via IPP to generate a PPD file for the
      printer */
-  response = get_printer_attributes(uri, NULL, 0, NULL, 0, 1);
+  std_uri = (char *)malloc(2048*(sizeof(char)));
```

Comment:
Hi @nidhijainpnp ,

please free any heap allocated memory before you exit the function if you don't return it as a parameter/return value.

Reply:
@nidhijainpnp, did you check whether all allocated memory gets correctly freed?

Reply:
yes